repo/open.go

```diff
@@ -333,6 +408,54 @@ func addThrottler(ctx context.Context, st blob.Storage) (blob.Storage, throttlin
 	return throttling.NewWrapper(st, throttler), throttler, nil
 }
 
+func upgradeLockMonitor(
+	ctx context.Context,
+	upgradeOwnerID string,
+	st blob.Storage,
+	password string,
+	cacheOpts *content.CachingOptions,
+	lockRefreshInterval time.Duration,
+	lastSync time.Time,
+	now func() time.Time,
+) blob.Storage {
+	var m sync.RWMutex
+
+	cb := func() error {
+		// protected read for lastSync because it will be shared between
+		// parallel storage operations
+		m.RLock()
+		if lastSync.Add(lockRefreshInterval).Before(now()) {
+			m.RUnlock()
+			return nil
+		}
+		m.RUnlock()
+
+		cacheTime, repoConfig, err := ReadAndCacheRepoConfig(ctx, st, password, cacheOpts, lockRefreshInterval)
+		if err != nil {
+			return err
+		}
+
+		// only allow the upgrade owner to perform storage operations
+		if locked, _ := repoConfig.UpgradeLock.IsLocked(now()); locked && upgradeOwnerID != repoConfig.UpgradeLock.OwnerID {
+			return ErrRepositoryUnavailableDueToUpgrageInProgress
+		}
+
+		m.Lock()
```

Comment:
shouldn't we move the `Lock()` right after `RUnlock()` to have full fence around caching?

Reply:
The desire is to avoid a mutex around a possible network bound operation to minimize the risk of locking down all goroutines. The caching mechanism uses atomic cache file operations, so this should not cause any corruption.

I understand that this might cause an inconsistent cache timestamp being tracked, but that'd be a minor compromise to favor a reduced scope for the lock.

Reply:
Will do.